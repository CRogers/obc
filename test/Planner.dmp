       |MODULE Planner;
       |
       |IMPORT In, Out, M := MathL, Conv, GC;
       |
       |TYPE R = LONGREAL;
       |
       |TYPE word = ARRAY 32 OF CHAR;
       |
       |(* Information about a town is stored as a structure with many links. *)
       |
       |  town = POINTER TO townrec;
       |  road = POINTER TO roadrec;
       |
       |  townrec = RECORD
       |    Tx, Ty: R;		(* X and Y coordinates *)
       |    Tdist: R;		(* Best known distance from start *)
       |    Troads: road;		(* List of roads starting here *)
       |    Thlink: town;		(* Next town in hash chain *)
       |    Tprev, Tnext: town;		(* Prev and Next towns in queue *)
       |    Tbacklink: town;		(* Prev town on shortest path *)
       |    Tname: word;		(* Name of the town *)
       |  END;
       |
       |  roadrec = RECORD
       |    Rdest: town;		(* Town reached by road *)
       |    Rlength: R;		(* Length of road *)
       |    Rnext: road;		(* Next road from same source *)
       |  END;    
       |
       |(* The towns live in a small hash table, so that a town can be found
       |   quickly, given its name. *)
       |
       |CONST HSIZE = 127;
       |
       |VAR hashtable: ARRAY HSIZE OF town;
       |
       |(* There's a priority queue of towns, ordered by best known distance
       |   from the origin.  We implement it as a doubly-linked ordered list. *)
       |
       |VAR qhead: town;		(* List head *)
       |
       |VAR
       |  hflag: BOOLEAN;		(* If true, use the Euclidean heuristic *)
       |  goal: town;			(* Goal town (used by heuristic) *)
       |
       |(* ---------------------------------------------------------------- *)
       |
       |PROCEDURE Panic(msg: ARRAY OF CHAR);
       |BEGIN
       |  Out.String(msg); Out.Ln;
       |  HALT(2)
       |END Panic;
       |
    148|PROCEDURE Split(VAR line: ARRAY OF CHAR; VAR words: ARRAY OF word): SHORTINT;
       |  VAR i, j: SHORTINT; n: INTEGER;
       |BEGIN
    148|  n := 0; i := 0;
       |  LOOP
   1184|    WHILE line[i] = ' ' DO INC(i) END;
    740|    IF line[i] = 0X THEN EXIT END;
    592|    IF line[i] = '{' THEN
      4|      INC(i); j := 0;
     52|      WHILE (line[i] # '}') & (line[i] # 0X) DO
     48|	words[n][j] := line[i];
     48|	INC(i); INC(j)
       |      END;
      4|      words[n][j] := 0X;
      4|      IF line[i] = '}' THEN INC(i) END
       |    ELSE
    588|      j := 0;
   4117|      WHILE (line[i] # ' ') & (line[i] # 0X) DO
   3529|	words[n][j] := line[i];
   3529|	INC(i); INC(j)
       |      END;
    588|      words[n][j] := 0X
       |    END;
    592|    INC(n)
       |  END;
    148|  RETURN SHORT(n)
       |END Split;
       |
       |(* euclid -- Euclidean distance between towns *)
     97|PROCEDURE Euclid(t, u: town): R;
       |  VAR dx, dy: R;
       |BEGIN
     97|  dx := t.Tx - u.Tx; dy := t.Ty - u.Ty;
       |  (* Out.LongReal(dx); Out.Ln;
       |  Out.LongReal(dy); Out.Ln; *)
     97|  RETURN M.Sqrt(dx*dx + dy*dy)
       |END Euclid;
       |
       |(* hash -- compute hash function of string *)
    246|PROCEDURE Hash(VAR name: word): INTEGER;
       |  VAR h: INTEGER; i: SHORTINT;
       |BEGIN
    246|  h := 0; i := 0;
   2330|  WHILE name[i] # 0X DO
   2084|    h := (5 * h + ORD(name[i])) MOD HSIZE;
   2084|    INC(i)
       |  END;
    246|  RETURN h
       |END Hash;
       |
       |(* lookup -- find a town in the hash table *)
    196|PROCEDURE Lookup(VAR name: word): town;
       |  VAR h: INTEGER; t: town;
       |BEGIN
    196|  h := Hash(name);
    196|  t := hashtable[h];
    237|  WHILE t # NIL DO
    237|    IF name = t.Tname THEN RETURN t END;
     41|    t := t.Thlink
       |  END;
       |  Panic("Unknown town");
       |  RETURN NIL
       |END Lookup;
       |
       |(* add_town -- add a new town to the hash table *)
     50|PROCEDURE AddTown(VAR name: word; x, y: R);
       |  VAR h: INTEGER; t: town;
       |BEGIN
     50|  h := Hash(name);
     50|  NEW(t); NEW(t);
     50|  t.Tname := name;
     50|  t.Tx := x; t.Ty := y;
     50|  t.Tdist := 0.0;
     50|  t.Troads := NIL;
     50|  t.Thlink := hashtable[h];
     50|  t.Tprev := NIL; t.Tnext := NIL;
     50|  t.Tbacklink := NIL;
     50|  hashtable[h] := t
       |END AddTown;
       |
       |(* add_road -- add a road and its converse *)
     97|PROCEDURE AddRoad(VAR name1, name2: word; leng: R);
       |  VAR t1, t2: town; r1, r2: road;
       |BEGIN
     97|  t1 := Lookup(name1); t2 := Lookup(name2);
     97|  NEW(r1); NEW(r1); NEW(r2);
       |
     97|  IF leng < Euclid(t1, t2) THEN
       |    Out.LongReal(leng); Out.Ln;
       |    Out.LongReal(Euclid(t1, t2)); Out.Ln;
       |    Panic("Road is too short")
       |  END;
       |
     97|  r1.Rdest := t2; r2.Rdest := t1;
     97|  r1.Rlength := leng; r2.Rlength := leng;
     97|  r1.Rnext := t1.Troads; t1.Troads := r1;
     97|  r2.Rnext := t2.Troads; t2.Troads := r2;
       |END AddRoad;
       |
       |(* ---------------------------------------------------------------- *)
       |
       |(* There's a priority queue of towns, ordered by best known distance
       |   from the origin.  We implement it as a doubly-linked ordered list. *)
       |
       |(* init_queue -- initialize the queue *)
      1|PROCEDURE InitQueue();
       |BEGIN
      1|  NEW(qhead);
      1|  qhead.Tdist := -1.0;
      1|  qhead.Tnext := qhead;
      1|  qhead.Tprev := qhead
       |END InitQueue;
       |
       |(* clear_queue -- set the queue to be empty *)
      1|PROCEDURE ClearQueue();
       |BEGIN
      1|  qhead.Tnext := qhead;
      1|  qhead.Tprev := qhead
       |END ClearQueue;
       |
       |(* enqueue -- add a town to the queue *)
     47|PROCEDURE Enqueue(t: town);
       |  VAR p, q: town;
       |BEGIN
       |     (* Since towns are visited in order of increasing distance, it
       |        seems reasonable to BEGIN the insertion PROCEDUREess from the tail
       |        of the queue *)
       |
     47|     p := qhead.Tprev;
    145|     WHILE t.Tdist < p.Tdist DO p := p.Tprev END;
     47|     q := p.Tnext;
     47|     t.Tprev := p; t.Tnext := q;
     47|     p.Tnext := t; q.Tprev := t;
       |END Enqueue;
       |
       |(* delmin -- delete and return the first town in the queue *)
     45|PROCEDURE Delmin(): town;
       |  VAR p, q: town;
       |BEGIN
     45|     p := qhead.Tnext;
     45|     IF p = qhead THEN RETURN NIL END;
     45|     q := p.Tnext;
     45|     qhead.Tnext := q; q.Tprev := qhead; 
     45|     RETURN p;
       |END Delmin;
       |
       |(* requeue -- a town has reduced in distance: adjust its place in the queue *)
      4|PROCEDURE Requeue(t: town);
       |  VAR p, q: town;
       |BEGIN
      4|     p := t.Tprev; q := t.Tnext;
      4|     p.Tnext := q; q.Tprev := p;
      5|     WHILE t.Tdist < p.Tdist DO p := p.Tprev END;
      4|     q := p.Tnext;
      4|     t.Tprev := p; t.Tnext := q;
      4|     p.Tnext := t; q.Tprev := t
       |END Requeue;
       |
       |(* ---------------------------------------------------------------- *)
       |
       |(* Now here's our implementation of Dijkstra's algorithm *)
       |
       |(* init -- initialize the search *)
      1|PROCEDURE Init(src, dst: town; heur: BOOLEAN);
       |  VAR i: INTEGER; t: town;
       |BEGIN
      1|  ClearQueue();
       |
    128|  FOR i := 0 TO HSIZE-1 DO
    127|    t := hashtable[i];
    177|    WHILE t # NIL DO
     50|      t.Tbacklink := NIL;
     50|      t := t.Thlink
       |    END
       |  END;         
       |
      1|  src.Tdist := 0.0;
      1|  src.Tbacklink := src;
      1|  goal := dst;
      1|  hflag := heur;
       |END Init;
       |
       |(* show_link -- send command to colour a road *)
    105|PROCEDURE ShowLink(t: town; colour: ARRAY OF CHAR);
       |BEGIN
    105|  Out.String("colour {"); Out.String(t.Tname); Out.String("} {");
    105|  Out.String(t.Tbacklink.Tname); Out.String("} ");
    105|  Out.String(colour); Out.Ln
       |END ShowLink;
       |
       |(* visit_neighbours -- update distances to neighbours of a town *)
     45|PROCEDURE VisitNeighbours(t: town);
       |  VAR r: road; u: town; d: R;
       |BEGIN
     45|  r := t.Troads;
    223|  WHILE r # NIL DO
    178|    u := r.Rdest;
    178|    d := t.Tdist + r.Rlength;
       |
       |    (* Heuristic: reduce the length of each road by the distance
       |       it takes us closer to the goal. *)
    178|    IF hflag THEN d := d - Euclid(t, goal) + Euclid(u, goal) END;
       |
    178|    IF u.Tbacklink = NIL THEN
       |      (* First time we've seen u *)
     47|      u.Tdist := d;
     47|      u.Tbacklink := t;
     47|      ShowLink(u, "green");
     47|      Enqueue(u);
    131|    ELSIF d < u.Tdist THEN
       |      (* Previous link to u has been improved *)
      4|      ShowLink(u, "brown");
      4|      u.Tdist := d;
      4|      u.Tbacklink := t;
      4|      ShowLink(u, "green");
      4|      Requeue(u)
       |    END;
       |
    178|    r := r.Rnext
       |  END
       |END VisitNeighbours;
       |
       |(* search -- main search function *)
      1|PROCEDURE Search(VAR sname, dname: word; heur: BOOLEAN);
       |  VAR t, src, dst: town; d: R;
       |BEGIN
      1|  src := Lookup(sname);
      1|  dst := Lookup(dname);
       |
      1|  Init(src, dst, heur);
      1|  t := src;
       |
      1|  GC.Collect;
       |
     46|  WHILE t # dst DO
     45|    IF t # src THEN
     44|      Out.String("paint {"); Out.String(t.Tname); 
     44|      Out.String("} white"); Out.Ln
       |    END;
     45|    VisitNeighbours(t);
     45|    Out.String("pause"); Out.Ln;
       |
     45|    t := Delmin();
     45|    IF t = NIL THEN
       |      Out.String("unreachable"); Out.Ln;
       |      RETURN
       |    END;
       |
     45|    ShowLink(t, "blue");
       |  END;
       |
      1|  Out.String("pause"); Out.Ln;
       |
       |  (* Show the shortest route in yellow *)
      1|  t := dst;
      6|  WHILE t # src DO
      5|    ShowLink(t, "yellow");
      5|    t := t.Tbacklink
       |  END;
       |
      1|  d := dst.Tdist;
      1|  IF hflag THEN d := d + Euclid(src, dst) END;
      1|  Out.String("found "); Out.Fixed(d, 0, 1); Out.Ln;
       |END Search;
       |
       |(* main -- main program *)
      1|PROCEDURE Main();
       |  VAR nwords: INTEGER;
       |    buf: ARRAY 256 OF CHAR;
       |    words: ARRAY 10 OF word;
       |BEGIN
      1|  InitQueue();
       |
       |  LOOP
    149|    In.Line(buf);
    149|    IF ~ In.Done THEN EXIT END;
       |    (* Out.String(buf); Out.Ln; *)
    148|    nwords := Split(buf, words);
       |
    148|    IF (nwords = 4) & (words[0] = "town") THEN
     50|      AddTown(words[1], Conv.RealVal(words[2]), Conv.RealVal(words[3]))
     98|    ELSIF (nwords = 4) & (words[0] = "road") THEN
     97|      AddRoad(words[1], words[2], Conv.RealVal(words[3]))
      1|    ELSIF (nwords = 4) & (words[0] = "search") THEN
      1|      Search(words[1], words[2], (words[3][0] = '1'))
       |    ELSE
       |      Panic("bad command")
       |    END
       |  END
       |END Main;
       |
       |BEGIN
      1|  Main()
       |END Planner.
