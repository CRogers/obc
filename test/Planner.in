town Aberdeen 244.199 500.825
town Aberystwyth 160.314 174.605
town Barnstaple 159.071 82.6424
town Birmingham 252.898 177.712
town Brighton 329.948 64.6226
town Bristol 223.072 106.876
town Cambridge 338.647 160.314
town Cardiff 197.596 109.361
town Carlisle 210.645 345.482
town Carmarthen 149.75 136.702
town Colchester 372.201 139.809
town Dorchester 229.286 55.9234
town Dover 392.707 87.6133
town Edinburgh 201.946 418.183
town Exeter 181.44 57.1661
town {Fort William} 130.488 480.941
town Glasgow 160.314 413.212
town Gloucester 237.985 135.459
town Guildford 310.064 92.5843
town Hereford 218.101 148.508
town Holyhead 139.187 237.364
town Hull 315.657 266.568
town Inverness 165.285 525.059
town Kendal 218.101 305.715
town Leeds 266.568 268.432
town Lincoln 308.821 230.529
town Liverpool 207.538 242.335
town Maidstone 357.288 96.3125
town Manchester 238.607 246.684
town Middlesborough 278.996 322.492
town Newcastle 263.461 350.453
town Northampton 295.151 161.557
town Norwich 387.114 191.382
town Nottingham 283.967 210.645
town Oxford 280.238 128.002
town Penzance 91.3416 18.6411
town Perth 193.246 449.251
town Peterborough 322.492 185.169
town Plymouth 153.479 33.554
town Portsmouth 288.316 62.1371
town Preston 219.344 266.568
town Salisbury 257.248 80.1569
town Sheffield 270.296 240.471
town Shrewsbury 216.859 193.868
town Southampton 274.025 69.5936
town Stoke-on-Trent 241.092 215.616
town Stranraer 127.381 347.968
town Taunton 200.082 77.05
town York 285.831 280.238
town London 328.084 111.225
road Penzance Plymouth 77
road Plymouth Exeter 45
road Exeter Taunton 35
road Taunton Salisbury 69
road Salisbury Southampton 23
road Southampton Portsmouth 21
road Portsmouth Guildford 47
road Guildford London 30
road London Oxford 56
road Oxford Gloucester 48
road Gloucester Birmingham 53
road Birmingham Stoke-on-Trent 48
road Stoke-on-Trent Sheffield 53
road Sheffield Manchester 38
road Manchester Preston 32
road Preston Kendal 43
road Kendal Carlisle 47
road Carlisle Newcastle 59
road Newcastle Edinburgh 108
road Edinburgh Glasgow 46
road Glasgow Perth 62
road Perth Aberdeen 86
road Aberdeen Inverness 106
road Inverness {Fort William} 65
road Taunton Bristol 52
road Bristol Cardiff 48
road Cardiff Carmarthen 68
road Carmarthen Aberystwyth 50
road Shrewsbury Birmingham 48
road Birmingham Nottingham 54
road Birmingham Northampton 55
road Gloucester Bristol 36
road Liverpool Manchester 35
road Liverpool Preston 36
road {Fort William} Perth 103
road Perth Edinburgh 42
road Newcastle Middlesborough 39
road Newcastle Leeds 97
road Leeds York 24
road Leeds Sheffield 36
road Sheffield Nottingham 44
road Nottingham Northampton 66
road Northampton London 68
road London Colchester 62
road London Cambridge 60
road Cambridge Peterborough 38
road Stranraer Carlisle 107
road Hull Leeds 60
road Hull York 38
road Oxford Northampton 43
road Oxford Southampton 66
road Stoke-on-Trent Manchester 45
road Carlisle Glasgow 97
road London Maidstone 39
road Maidstone Dover 41
road London Brighton 59
road Brighton Portsmouth 49
road Hereford Gloucester 31
road Hereford Shrewsbury 52
road Norwich Colchester 59
road Norwich Peterborough 77
road Norwich Cambridge 63
road Peterborough Lincoln 50
road Peterborough Nottingham 58
road Nottingham Lincoln 36
road Lincoln Hull 48
road York Middlesborough 50
road Dorchester Southampton 54
road Dorchester Salisbury 40
road Dorchester Taunton 45
road Dorchester Exeter 55
road Barnstaple Plymouth 61
road Barnstaple Exeter 55
road Barnstaple Taunton 50
road Holyhead Liverpool 102
road Aberystwyth Hereford 79
road Stranraer Glasgow 85
road Holyhead Shrewsbury 104
road Aberystwyth Shrewsbury 74
road Shrewsbury Liverpool 66
road Shrewsbury Stoke-on-Trent 53
road Holyhead Aberystwyth 105
road Inverness Perth 114
road Middlesborough Leeds 64
road Kendal Leeds 72
road Kendal Newcastle 92
road Preston Leeds 68
road Bristol Salisbury 54
road Bristol Oxford 74
road Glasgow {Fort William} 102
road Brighton Dover 81
road Brighton Maidstone 66
road Northampton Cambridge 54
road Cambridge Colchester 48
road Sheffield Lincoln 47
road Sheffield Hull 66
road Oxford Birmingham 68
search Birmingham Edinburgh 0
