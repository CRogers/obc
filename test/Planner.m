MODULE Planner;

IMPORT In, Out, M := MathL, Conv, GC;

TYPE R = LONGREAL;

TYPE word = ARRAY 32 OF CHAR;

(* Information about a town is stored as a structure with many links. *)

  town = POINTER TO townrec;
  road = POINTER TO roadrec;

  townrec = RECORD
    Tx, Ty: R;		(* X and Y coordinates *)
    Tdist: R;		(* Best known distance from start *)
    Troads: road;		(* List of roads starting here *)
    Thlink: town;		(* Next town in hash chain *)
    Tprev, Tnext: town;		(* Prev and Next towns in queue *)
    Tbacklink: town;		(* Prev town on shortest path *)
    Tname: word;		(* Name of the town *)
  END;

  roadrec = RECORD
    Rdest: town;		(* Town reached by road *)
    Rlength: R;		(* Length of road *)
    Rnext: road;		(* Next road from same source *)
  END;    

(* The towns live in a small hash table, so that a town can be found
   quickly, given its name. *)

CONST HSIZE = 127;

VAR hashtable: ARRAY HSIZE OF town;

(* There's a priority queue of towns, ordered by best known distance
   from the origin.  We implement it as a doubly-linked ordered list. *)

VAR qhead: town;		(* List head *)

VAR
  hflag: BOOLEAN;		(* If true, use the Euclidean heuristic *)
  goal: town;			(* Goal town (used by heuristic) *)

(* ---------------------------------------------------------------- *)

PROCEDURE Panic(msg: ARRAY OF CHAR);
BEGIN
  Out.String(msg); Out.Ln;
  HALT(2)
END Panic;

PROCEDURE Split(VAR line: ARRAY OF CHAR; VAR words: ARRAY OF word): SHORTINT;
  VAR i, j: SHORTINT; n: INTEGER;
BEGIN
  n := 0; i := 0;
  LOOP
    WHILE line[i] = ' ' DO INC(i) END;
    IF line[i] = 0X THEN EXIT END;
    IF line[i] = '{' THEN
      INC(i); j := 0;
      WHILE (line[i] # '}') & (line[i] # 0X) DO
	words[n][j] := line[i];
	INC(i); INC(j)
      END;
      words[n][j] := 0X;
      IF line[i] = '}' THEN INC(i) END
    ELSE
      j := 0;
      WHILE (line[i] # ' ') & (line[i] # 0X) DO
	words[n][j] := line[i];
	INC(i); INC(j)
      END;
      words[n][j] := 0X
    END;
    INC(n)
  END;
  RETURN SHORT(n)
END Split;

(* euclid -- Euclidean distance between towns *)
PROCEDURE Euclid(t, u: town): R;
  VAR dx, dy: R;
BEGIN
  dx := t.Tx - u.Tx; dy := t.Ty - u.Ty;
  (* Out.LongReal(dx); Out.Ln;
  Out.LongReal(dy); Out.Ln; *)
  RETURN M.Sqrt(dx*dx + dy*dy)
END Euclid;

(* hash -- compute hash function of string *)
PROCEDURE Hash(VAR name: word): INTEGER;
  VAR h: INTEGER; i: SHORTINT;
BEGIN
  h := 0; i := 0;
  WHILE name[i] # 0X DO
    h := (5 * h + ORD(name[i])) MOD HSIZE;
    INC(i)
  END;
  RETURN h
END Hash;

(* lookup -- find a town in the hash table *)
PROCEDURE Lookup(VAR name: word): town;
  VAR h: INTEGER; t: town;
BEGIN
  h := Hash(name);
  t := hashtable[h];
  WHILE t # NIL DO
    IF name = t.Tname THEN RETURN t END;
    t := t.Thlink
  END;
  Panic("Unknown town");
  RETURN NIL
END Lookup;

(* add_town -- add a new town to the hash table *)
PROCEDURE AddTown(VAR name: word; x, y: R);
  VAR h: INTEGER; t: town;
BEGIN
  h := Hash(name);
  NEW(t); NEW(t);
  t.Tname := name;
  t.Tx := x; t.Ty := y;
  t.Tdist := 0.0;
  t.Troads := NIL;
  t.Thlink := hashtable[h];
  t.Tprev := NIL; t.Tnext := NIL;
  t.Tbacklink := NIL;
  hashtable[h] := t
END AddTown;

(* add_road -- add a road and its converse *)
PROCEDURE AddRoad(VAR name1, name2: word; leng: R);
  VAR t1, t2: town; r1, r2: road;
BEGIN
  t1 := Lookup(name1); t2 := Lookup(name2);
  NEW(r1); NEW(r1); NEW(r2);

  IF leng < Euclid(t1, t2) THEN
    Out.LongReal(leng); Out.Ln;
    Out.LongReal(Euclid(t1, t2)); Out.Ln;
    Panic("Road is too short")
  END;

  r1.Rdest := t2; r2.Rdest := t1;
  r1.Rlength := leng; r2.Rlength := leng;
  r1.Rnext := t1.Troads; t1.Troads := r1;
  r2.Rnext := t2.Troads; t2.Troads := r2;
END AddRoad;

(* ---------------------------------------------------------------- *)

(* There's a priority queue of towns, ordered by best known distance
   from the origin.  We implement it as a doubly-linked ordered list. *)

(* init_queue -- initialize the queue *)
PROCEDURE InitQueue();
BEGIN
  NEW(qhead);
  qhead.Tdist := -1.0;
  qhead.Tnext := qhead;
  qhead.Tprev := qhead
END InitQueue;

(* clear_queue -- set the queue to be empty *)
PROCEDURE ClearQueue();
BEGIN
  qhead.Tnext := qhead;
  qhead.Tprev := qhead
END ClearQueue;

(* enqueue -- add a town to the queue *)
PROCEDURE Enqueue(t: town);
  VAR p, q: town;
BEGIN
     (* Since towns are visited in order of increasing distance, it
        seems reasonable to BEGIN the insertion PROCEDUREess from the tail
        of the queue *)

     p := qhead.Tprev;
     WHILE t.Tdist < p.Tdist DO p := p.Tprev END;
     q := p.Tnext;
     t.Tprev := p; t.Tnext := q;
     p.Tnext := t; q.Tprev := t;
END Enqueue;

(* delmin -- delete and return the first town in the queue *)
PROCEDURE Delmin(): town;
  VAR p, q: town;
BEGIN
     p := qhead.Tnext;
     IF p = qhead THEN RETURN NIL END;
     q := p.Tnext;
     qhead.Tnext := q; q.Tprev := qhead; 
     RETURN p;
END Delmin;

(* requeue -- a town has reduced in distance: adjust its place in the queue *)
PROCEDURE Requeue(t: town);
  VAR p, q: town;
BEGIN
     p := t.Tprev; q := t.Tnext;
     p.Tnext := q; q.Tprev := p;
     WHILE t.Tdist < p.Tdist DO p := p.Tprev END;
     q := p.Tnext;
     t.Tprev := p; t.Tnext := q;
     p.Tnext := t; q.Tprev := t
END Requeue;

(* ---------------------------------------------------------------- *)

(* Now here's our implementation of Dijkstra's algorithm *)

(* init -- initialize the search *)
PROCEDURE Init(src, dst: town; heur: BOOLEAN);
  VAR i: INTEGER; t: town;
BEGIN
  ClearQueue();

  FOR i := 0 TO HSIZE-1 DO
    t := hashtable[i];
    WHILE t # NIL DO
      t.Tbacklink := NIL;
      t := t.Thlink
    END
  END;         

  src.Tdist := 0.0;
  src.Tbacklink := src;
  goal := dst;
  hflag := heur;
END Init;

(* show_link -- send command to colour a road *)
PROCEDURE ShowLink(t: town; colour: ARRAY OF CHAR);
BEGIN
  Out.String("colour {"); Out.String(t.Tname); Out.String("} {");
  Out.String(t.Tbacklink.Tname); Out.String("} ");
  Out.String(colour); Out.Ln
END ShowLink;

(* visit_neighbours -- update distances to neighbours of a town *)
PROCEDURE VisitNeighbours(t: town);
  VAR r: road; u: town; d: R;
BEGIN
  r := t.Troads;
  WHILE r # NIL DO
    u := r.Rdest;
    d := t.Tdist + r.Rlength;

    (* Heuristic: reduce the length of each road by the distance
       it takes us closer to the goal. *)
    IF hflag THEN d := d - Euclid(t, goal) + Euclid(u, goal) END;

    IF u.Tbacklink = NIL THEN
      (* First time we've seen u *)
      u.Tdist := d;
      u.Tbacklink := t;
      ShowLink(u, "green");
      Enqueue(u);
    ELSIF d < u.Tdist THEN
      (* Previous link to u has been improved *)
      ShowLink(u, "brown");
      u.Tdist := d;
      u.Tbacklink := t;
      ShowLink(u, "green");
      Requeue(u)
    END;

    r := r.Rnext
  END
END VisitNeighbours;

(* search -- main search function *)
PROCEDURE Search(VAR sname, dname: word; heur: BOOLEAN);
  VAR t, src, dst: town; d: R;
BEGIN
  src := Lookup(sname);
  dst := Lookup(dname);

  Init(src, dst, heur);
  t := src;

  GC.Collect;

  WHILE t # dst DO
    IF t # src THEN
      Out.String("paint {"); Out.String(t.Tname); 
      Out.String("} white"); Out.Ln
    END;
    VisitNeighbours(t);
    Out.String("pause"); Out.Ln;

    t := Delmin();
    IF t = NIL THEN
      Out.String("unreachable"); Out.Ln;
      RETURN
    END;

    ShowLink(t, "blue");
  END;

  Out.String("pause"); Out.Ln;

  (* Show the shortest route in yellow *)
  t := dst;
  WHILE t # src DO
    ShowLink(t, "yellow");
    t := t.Tbacklink
  END;

  d := dst.Tdist;
  IF hflag THEN d := d + Euclid(src, dst) END;
  Out.String("found "); Out.Fixed(d, 0, 1); Out.Ln;
END Search;

(* main -- main program *)
PROCEDURE Main();
  VAR nwords: INTEGER;
    buf: ARRAY 256 OF CHAR;
    words: ARRAY 10 OF word;
BEGIN
  InitQueue();

  LOOP
    In.Line(buf);
    IF ~ In.Done THEN EXIT END;
    (* Out.String(buf); Out.Ln; *)
    nwords := Split(buf, words);

    IF (nwords = 4) & (words[0] = "town") THEN
      AddTown(words[1], Conv.RealVal(words[2]), Conv.RealVal(words[3]))
    ELSIF (nwords = 4) & (words[0] = "road") THEN
      AddRoad(words[1], words[2], Conv.RealVal(words[3]))
    ELSIF (nwords = 4) & (words[0] = "search") THEN
      Search(words[1], words[2], (words[3][0] = '1'))
    ELSE
      Panic("bad command")
    END
  END
END Main;

BEGIN
  Main()
END Planner.
