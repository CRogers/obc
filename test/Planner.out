colour {Oxford} {Birmingham} green
colour {Northampton} {Birmingham} green
colour {Nottingham} {Birmingham} green
colour {Shrewsbury} {Birmingham} green
colour {Stoke-on-Trent} {Birmingham} green
colour {Gloucester} {Birmingham} green
pause
colour {Shrewsbury} {Birmingham} blue
paint {Shrewsbury} white
colour {Liverpool} {Shrewsbury} green
colour {Aberystwyth} {Shrewsbury} green
colour {Holyhead} {Shrewsbury} green
colour {Hereford} {Shrewsbury} green
pause
colour {Stoke-on-Trent} {Birmingham} blue
paint {Stoke-on-Trent} white
colour {Manchester} {Stoke-on-Trent} green
colour {Sheffield} {Stoke-on-Trent} green
pause
colour {Gloucester} {Birmingham} blue
paint {Gloucester} white
colour {Hereford} {Shrewsbury} brown
colour {Hereford} {Gloucester} green
colour {Bristol} {Gloucester} green
pause
colour {Nottingham} {Birmingham} blue
paint {Nottingham} white
colour {Lincoln} {Nottingham} green
colour {Peterborough} {Nottingham} green
colour {Sheffield} {Stoke-on-Trent} brown
colour {Sheffield} {Nottingham} green
pause
colour {Northampton} {Birmingham} blue
paint {Northampton} white
colour {Cambridge} {Northampton} green
colour {London} {Northampton} green
pause
colour {Oxford} {Birmingham} blue
paint {Oxford} white
colour {Southampton} {Oxford} green
pause
colour {Hereford} {Gloucester} blue
paint {Hereford} white
pause
colour {Bristol} {Gloucester} blue
paint {Bristol} white
colour {Salisbury} {Bristol} green
colour {Cardiff} {Bristol} green
colour {Taunton} {Bristol} green
pause
colour {Lincoln} {Nottingham} blue
paint {Lincoln} white
colour {Hull} {Lincoln} green
pause
colour {Manchester} {Stoke-on-Trent} blue
paint {Manchester} white
colour {Preston} {Manchester} green
pause
colour {Sheffield} {Nottingham} blue
paint {Sheffield} white
colour {Leeds} {Sheffield} green
pause
colour {Cambridge} {Northampton} blue
paint {Cambridge} white
colour {Colchester} {Cambridge} green
colour {Norwich} {Cambridge} green
pause
colour {Peterborough} {Nottingham} blue
paint {Peterborough} white
pause
colour {Liverpool} {Shrewsbury} blue
paint {Liverpool} white
pause
colour {Aberystwyth} {Shrewsbury} blue
paint {Aberystwyth} white
colour {Carmarthen} {Aberystwyth} green
pause
colour {London} {Northampton} blue
paint {London} white
colour {Brighton} {London} green
colour {Maidstone} {London} green
colour {Guildford} {London} green
pause
colour {Preston} {Manchester} blue
paint {Preston} white
colour {Kendal} {Preston} green
pause
colour {Southampton} {Oxford} blue
paint {Southampton} white
colour {Dorchester} {Southampton} green
colour {Portsmouth} {Southampton} green
pause
colour {Leeds} {Sheffield} blue
paint {Leeds} white
colour {Middlesborough} {Leeds} green
colour {York} {Leeds} green
colour {Newcastle} {Leeds} green
pause
colour {Cardiff} {Bristol} blue
paint {Cardiff} white
pause
colour {Hull} {Lincoln} blue
paint {Hull} white
pause
colour {Taunton} {Bristol} blue
paint {Taunton} white
colour {Barnstaple} {Taunton} green
colour {Dorchester} {Southampton} brown
colour {Dorchester} {Taunton} green
colour {Exeter} {Taunton} green
pause
colour {Salisbury} {Bristol} blue
paint {Salisbury} white
colour {Dorchester} {Taunton} brown
colour {Dorchester} {Salisbury} green
pause
colour {Holyhead} {Shrewsbury} blue
paint {Holyhead} white
pause
colour {Guildford} {London} blue
paint {Guildford} white
pause
colour {Portsmouth} {Southampton} blue
paint {Portsmouth} white
pause
colour {Colchester} {Cambridge} blue
paint {Colchester} white
pause
colour {York} {Leeds} blue
paint {York} white
pause
colour {Maidstone} {London} blue
paint {Maidstone} white
colour {Dover} {Maidstone} green
pause
colour {Kendal} {Preston} blue
paint {Kendal} white
colour {Carlisle} {Kendal} green
pause
colour {Norwich} {Cambridge} blue
paint {Norwich} white
pause
colour {Carmarthen} {Aberystwyth} blue
paint {Carmarthen} white
pause
colour {Exeter} {Taunton} blue
paint {Exeter} white
colour {Plymouth} {Exeter} green
pause
colour {Brighton} {London} blue
paint {Brighton} white
pause
colour {Dorchester} {Salisbury} blue
paint {Dorchester} white
pause
colour {Barnstaple} {Taunton} blue
paint {Barnstaple} white
pause
colour {Middlesborough} {Leeds} blue
paint {Middlesborough} white
pause
colour {Dover} {Maidstone} blue
paint {Dover} white
pause
colour {Carlisle} {Kendal} blue
paint {Carlisle} white
colour {Glasgow} {Carlisle} green
colour {Stranraer} {Carlisle} green
pause
colour {Plymouth} {Exeter} blue
paint {Plymouth} white
colour {Penzance} {Plymouth} green
pause
colour {Newcastle} {Leeds} blue
paint {Newcastle} white
colour {Edinburgh} {Newcastle} green
pause
colour {Penzance} {Plymouth} blue
paint {Penzance} white
pause
colour {Glasgow} {Carlisle} blue
paint {Glasgow} white
colour {Fort William} {Glasgow} green
colour {Perth} {Glasgow} green
pause
colour {Stranraer} {Carlisle} blue
paint {Stranraer} white
pause
colour {Edinburgh} {Newcastle} blue
pause
colour {Edinburgh} {Newcastle} yellow
colour {Newcastle} {Leeds} yellow
colour {Leeds} {Sheffield} yellow
colour {Sheffield} {Nottingham} yellow
colour {Nottingham} {Birmingham} yellow
found 339.0
